lib/common-utils.ts

```diff
@@ -258,3 +258,99 @@ export const isImageBufferData = (imageData: ImageBuffer | ImageFile | ImageBase
 export const isImageInfoWithState = (imageInfo: ImageInfoFull): imageInfo is ImageInfoWithState => {
     return Boolean((imageInfo as ImageInfoWithState).stateName);
 };
+
+export const trimArray = <T>(array: Array<T>): Array<T> => {
+    let indexBegin = 0;
+    let indexEnd = array.length;
+
+    while (indexBegin < array.length && !array[indexBegin]) {
+        indexBegin++;
+    }
+
+    while (indexEnd > 0 && !array[indexEnd - 1]) {
+        indexEnd--;
+    }
+
+    return array.slice(indexBegin, indexEnd);
+};
+
+const getErrorTitle = (e: Error): string => {
+    let errorName = e.name;
+
+    if (!errorName && e.stack) {
+        const columnIndex = e.stack.indexOf(':');
+
+        if (columnIndex !== -1) {
+            errorName = e.stack.slice(0, columnIndex);
+        } else {
+            errorName = e.stack.slice(0, e.stack.indexOf('\n'));
+        }
+    }
+
+    if (!errorName) {
+        errorName = 'Error';
+    }
+
+    return e.message ? `${errorName}: ${e.message}` : errorName;
+};
+
+const getErrorRawStackFrames = (e: Error & { stack: string }): string => {
+    const errorTitle = getErrorTitle(e) + '\n';
+    const errorTitleStackIndex = e.stack.indexOf(errorTitle);
+
+    if (errorTitleStackIndex !== -1) {
+        return e.stack.slice(errorTitleStackIndex + errorTitle.length);
+    }
+
+    const errorMessageStackIndex = e.stack.indexOf(e.message);
+    const errorMessageEndsStackIndex = e.stack.indexOf('\n', errorMessageStackIndex + e.message.length);
+
+    return e.stack.slice(errorMessageEndsStackIndex);
+};
+
+const cloneError = <T extends Error>(error: T): T => {
+    const originalProperties = ['name', 'message', 'stack'] as Array<keyof Error>;
+    const clonedError = new Error(error.message) as T;
+
+    originalProperties.forEach(property => {
+        delete clonedError[property];
+    });
+
+    const customProperties = Object.getOwnPropertyNames(error) as Array<keyof Error>;
+
+    originalProperties.concat(customProperties).forEach((property) => {
+        clonedError[property] = error[property] as any; // eslint-disable-line @typescript-eslint/no-explicit-any
+    });
+
+    return clonedError;
+};
+
+export const mergeSnippetIntoErrorStack = <T extends Error>(error: T & { snippet?: string }): T => {
+    if (!error.snippet) {
+        return error;
+    }
+
+    const clonedError = cloneError(error);
+
+    delete clonedError.snippet;
+
+    if (!error.stack) {
+        clonedError.stack = [
+            getErrorTitle(error),
+            error.snippet
+        ].join('\n');
+
+        return clonedError;
+    }
+
+    const grayBegin = '\x1B[90m';
+    const grayEnd = '\x1B[39m';
+
+    clonedError.stack = [
```

Comment:
it looks like duplicated code. it can be rewritten as

```javascript
const mergedStack = [getErrorTitle(error), error.snippet];

if (error.stack) {
    const grayBegin = '\x1B[90m';
    const grayEnd = '\x1B[39m';

    mergedStack.push(
        grayBegin + getErrorRawStackFrames(error as Error & { stack: string }) + grayEnd
    );
}

return mergedStack.join('\n');
```

also, we can add `colorCode` function for more readable code:

```javascript
function wrapWithAnsiColor(code, color = '90m') {
    return `\x1B[${color}${code}\x1B[39m`;
}
//...
if (error.stack) {
    const rawStackFrames = getErrorRawStackFrames(error as Error & { stack: string });

    mergedStack.push(
        wrapWithAnsiColor(rawStackFrames)
    );
}
```

Reply:
jfyi: `m` is not a part of the color code in ansi.

Exampels:
- `\033[9;31;40m` (`[9` - Strikethrough, `31` - text color, `40` - bg color): strike-through red text with black bg
- `\033[1;37;42m` (`[1` - Semibold, `37` - text color, `42` - bg color): semibold white text with green bg
- `\033[0m` (`[0` - normal text, ... - reset text color, ... - reset bg color): removes all styling